Add blocknotify and walletnotify functional tests

promag opened the pull request:
This patch adds the missing functional tests for `-blocknotify` and `-walletnotify` notifications. The `-alertnotify` test file `forknotify.py` is renamed to `notifications.py` to accommodate the new tests. Credits to @jnewbery for this cleanup and unification.

promag commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,42 @@
+    def run_test(self):
+        # generate 5 blocks
+        blocks = self.nodes[0].generate(5)
+        sleep(1)
```
Without these sometimes the file doesn't have all blocks because `runCommand()` is aync. Any idea to avoid the sleep without pooling or blocking reads?

practicalswift commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,42 @@
+
+Verify that a bitcoind node calls the given command for each block
+"""
+from os import close, remove
```
Replace with `import os`. `os.close` and `os.remove` is used below :-)

promag replied:
Lol yes. What do you guys prefer? Import module or symbols?

practicalswift replied:
We have a winner:

```
$ git grep -E "(import os|from os)" -- "*.py" | cut -f2 -d: | sort | uniq -c
     25 import os
      2 import os.path
```

practicalswift commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,42 @@
+from os import close, remove
+from tempfile import mkstemp
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, assert_raises_jsonrpc
```
Remove `assert_raises_jsonrpc`. Not used :-)

practicalswift commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,42 @@
+        super().__init__()
+
+        # get a temporary file path to store the notified blocks
+        (fd, path) = mkstemp(text = True)
```
Nit: `fd, path = mkstemp(text = True)` will do :-)

practicalswift commented:
Concept ACK :-)

promag commented:
@practicalswift thanks for the review. Added a few comments and fixed the nits.

MarcoFalke commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,49 @@
+        super().__init__()
+
+        # get a temporary file path to store the notified blocks
+        fd, path = tempfile.mkstemp(text = True)
```
Please use `self.options.tmpdir`, so that files are contained in a single tree for each test.

Also nit: Please no whitespace for named args. (`text=True`)

promag replied:
`self.options` is not defined at this point.

MarcoFalke commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,49 @@
+        blocks = self.nodes[0].generate(3)
+
+        # wait while the command runs asynchronously
+        time.sleep(3)
```
Please no fixed sleeps. This might be too much time on a fast system or too less time on a slow io system.

You could wrap it into a while loop with timeout on failure and break on success. Alternatively there should be some helper function `wait_until` if I am not mistaken.

maflcko reviewed (commented):
Concept ACK, but needs travis fixed:

```
WARNING! The following scripts are not being run: ['blocknotify.py']. Check the test lists in test_runner.py.

jnewbery commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,49 @@
+        # file content should equal the generated blocks hashes
+        with open(self.path, 'r') as f:
+            assert_equal(''.join(blocks), f.read())
+        
```
nit: whitespace.

Consider using a linter to catch nits.

jnewbery commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,49 @@
+        fd, path = tempfile.mkstemp(text = True)
+        os.close(fd)
+
+        self.path = path
```
can assign `self.path` above rather than assign to a local variable `path`

jnewbery reviewed (commented):
A couple of nits to add to @MarcoFalke's

You'll need to add `blocknotify.py` to `BASE_SCRIPTS` in test_runner.py

promag commented:
@MarcoFalke done. However can't get tmpdir in the constructor. Any other suggestion?

MarcoFalke commented:
You could try to overwrite setup_network:

```py
def setup_network():
    self.extra_args = [bla]
    super().setup_network()
```

On Mon, Jul 31, 2017 at 10:06 PM, João Barbosa <notifications@github.com>
wrote:

> @MarcoFalke <https://github.com/marcofalke> done. However can't get
> tmpdir in the constructor. Any other suggestion?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/10941#issuecomment-319180756>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGGmv0EceWaRTPKc4AFnANOaqKf4O_Jqks5sTjOogaJpZM4OlIuV>
> .
>

promag commented:
@MarcoFalke ty, all comments addressed.

promag committed:
[tests] Add -blocknotify functional test

promag committed:
[tests] Add -walletnotify functional test

practicalswift commented:
utACK d2c730c77f9b4011040e30f2bc2a4c8c6deb13f2

jnewbery commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,47 @@
+        block_count = 10
+        blocks = self.nodes[0].generate(block_count)
+
+        # wait for expected data length
```
A few comments on this loop:

- I'd make the sleep longer (say 0.1 seconds). I'm also not sure you need the back-off. Just a `time.sleep(0.1)` should be fine.
- The total time allowed for the file to be written is 0.45 seconds. That needs to be longer (up to 10 seconds) since Travis can block i/o for many seconds at a time. See #10072.
- Please make the comment clear about the total time allowed for the loop.

jnewbery reviewed (commented):
tested ACK with one change needed to the timing of the while loop

promag commented:
@jnewbery done. Also rebased.

MarcoFalke commented:
Needs rebase

promag commented:
@MarcoFalke done.

MarcoFalke commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,47 @@
+
+        # wait at most 10 seconds for expected file size before reading the content
+        attempts = 0
+        while os.stat(self.path).st_size < (block_count * 65) and attempts < 100:
```
Maybe wait until the file is created? I'd really suggest using some sort of wrapper, like the `wait_until` from `mininode`.

```py
wait_until(lambda: os.path.isfile(self.path), timeout=1)
wait_until(lambda: os.stat(self.path).st_size >= (block_count * 65), timeout=10)

MarcoFalke replied:
ping @promag: Are you still working on this?

ryanofsky replied:
> ping @promag: Are you still working on this?

I'm not seeing any errors when I run the test, but this is probably caused by a race between the start of this loop and the first notification. Could be fixed by adding a `not os.path.exists(self.path)` check here, or adding `with open(self.path, "w"): pass` to `setup_network`.

MarcoFalke replied:
Indeed. The code I posted a few comments above fixed it for me.

c.f.

```py
wait_until(lambda: os.path.isfile(self.path), timeout=1)
wait_until(lambda: os.stat(self.path).st_size >= (block_count * 65), timeout=10)

jnewbery commented:
I still believe that this test should be merged with #10966 and the `forknotify.py` test. They're testing very similar functionality and there's a lot of shared code between the three tests.

MarcoFalke commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,47 @@
+
+        # wait at most 10 seconds for expected file size before reading the content
+        attempts = 0
+        while os.stat(self.path).st_size < (block_count * 65) and attempts < 100:
```
Just to make sure, that this is causing failures on my machine:

```py
  File "./test/functional/blocknotify.py", line 38, in run_test
    while os.stat(self.path).st_size < (block_count * 65) and attempts < 100:
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/bitcoin_test_runner_20170815_171120/blocknotify_621/blocks.txt'

ryanofsky commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,47 @@
+        self.path = os.path.join(self.options.tmpdir, 'blocks.txt')
+
+        # the command will append the notified block hash to file
+        self.extra_args = [['-blocknotify=echo %%s >> %s' % self.path]]
```
Maybe escape path with `shlex.quote(self.path)`

promag replied:
`tmpdir` is used elsewhere (should be secure) and `blocks.txt` doesn't need to be quoted. Is there a strong argument for it?

promag replied:
Of you mean to quote the command `echo %%s >> %s`?

ryanofsky replied:
Quoting the entire command would be incorrect because it is passed literally to [system(3)](http://man7.org/linux/man-pages/man3/system.3.html). Quoting the path is correct because it is being passed as a shell token. Anyway just suggested this in case you wanted the code to be more correct. There are no practical implications (security or otherwise) either way.

promag replied:
Right, otherwise it wouldn't work 😃 I'm going for simplicity if you don't mind.

ryanofsky reviewed (commented):
Tested ACK 2b03bc1fb844980ffc67c6eaa52b034e978a9826 as long as the race condition is fixed. 

I agree with others the test setup should be merged with  walletnotify test setup in #10966 so code is not duplicated, but I don't think it matters if this is done in one or two prs.

promag commented:
As suggested I'll merge the test and use `waut_until`.

promag commented:
Rebased and updated as per suggestions. Still in a separate test file because IMO that should go to a different PR since `forknotify.py` needs to be changed.

jnewbery committed:
[tests] Tidy up forknotify.py

jnewbery commented:
Please see https://github.com/jnewbery/bitcoin/tree/fork_notify. This adds blocknotify and walletnotify to the existint fork_notify.py test script.

MarcoFalke commented on test/functional/blocknotify.py:
```diff
@@ -0,0 +1,40 @@
+
+        # file content should equal the generated blocks hashes
+        with open(self.path, 'r') as f:
+            assert_equal(blocks, f.read().splitlines())
```
Hmm. I don't think boost::thread guarantees that the thread run in the order they are constructed. Am I wrong?

The arrays should have the same content, but a different order.

promag replied:
Correct, in `-walletnotify` (#10966) there's:
```cpp
            # txids are sorted because there is no order guarantee in notifications
            assert_equal(sorted(txids_rpc), sorted(txids_file))
```
So I'll do the same here for correctness.

maflcko reviewed (commented):
utACK e7408a444 but needs clarification

promag commented:
Picked your branch and added a commit that adds `-rescan` to `node[1]` (needs squash).

@MarcoFalke the block sort was already on the @jnewbery branch.

jnewbery commented:
Tested ACK 6d9c1593497e9f295b5f335c7c7a0df7966d7590. As you've pointed out, should be squashed prior to merge.

promag commented:
@jnewbery Done.

mess110 commented:
Is the name of the test class still correct? I think it should be something like `NotificationsTest`

mess110 commented on test/functional/notifications.py:
```diff
@@ -0,0 +1,86 @@
+# Copyright (c) 2014-2016 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the -alertnotify and -blocknotify options."""
```
You are also testing -walletnotify, could you please update the comment?

mess110 commented on test/functional/notifications.py:
```diff
@@ -0,0 +1,86 @@
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, wait_until, connect_nodes_bi
+
+class ForkNotifyTest(BitcoinTestFramework):
```
Is the class name still correct? I think it should be something like `NotificationsTest`

promag commented:
@mess110 Done, thank you.

mess110 commented:
ACK https://github.com/bitcoin/bitcoin/pull/10941/commits/857b32b4b280f13997cf2fa471802ad6a27075fb

(or the squashed version of this)

promag commented:
I think makes sense to have individual commits.

laanwj commented:
> I think makes sense to have individual commits.

Certainly if multiple authors are involved. And the different commits concern different aspects.

What we want to avoid with squashing is sequential commits like "fix the previous commit" which are internal to the author's working and are only confusing when reading the history. But that's not the case here.
